### DISCUSSION TITLE
Registering a custom IDGenerator

### DISCUSSION THREAD

**Discussion by gpalyan** [open-telemetry/opentelemetry-java-instrumentation#4858] [Q&A]
Hello,

I am trying to use your application within a Spring Boot app. I am trying to incorporate a custom IDGenerator but can't seem to get it to work. This is a sample code I have:

Configuration of SDK:

@Configuration
public class CustomSdkTracerProviderConfiguration {

    @Bean
    public SdkTracerProvider sdkTracerProvider() {
        return SdkTracerProvider.builder().setIdGenerator(new MyCustomIdGenerator()).build();
    }

    @Bean
    public OpenTelemetrySdk openTelemetrySdk(@NonNull SdkTracerProvider sdkTracerProvider) {
        OpenTelemetrySdk openTelemetrySdk =
                OpenTelemetrySdk.builder()
                        .setTracerProvider(sdkTracerProvider)
                        .buildAndRegisterGlobal();
        return openTelemetrySdk;
    }
}

Usage:
Span.current().getSpanContext().getTraceId();

Is there something I am missing here? I wasn't able to find great examples of how to register custom classes into your SDK.

**Comment by jkwatson** [comment]
Are you also using the otel javaagent? Or is this an SDK-only application?

**Reply by gpalyan** [reply]
Yes I am using the javaagent. Not certain if that's the best approach or not. Can you advise?

**Reply by jkwatson** [reply]
I'm going to move this question over to the instrumentation repo, because it is not trivial and I'm not sure how to do it myself.

**Reply by gpalyan** [reply]
Thanks!

**Reply by trask** [reply]
hi @gpalyan! you'll need to build an "agent extension", see https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension#introduction and https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoIdGenerator.java

**Reply by gpalyan** [reply]
Can you clarify wht's needed in the extension jar? Do I just put the custom IDGenerator and it will work? Or is there more to it?

**Reply by trask** [reply]
there's a bit more to it, it relies on Java SPI, see if you can get the extension example working as a first step

**Reply by gpalyan** [reply]
I tried to build this project and see if I can run it alongside the telemetry agent but I get following error.  Can you advise? 


* What went wrong:
Execution failed for task ':benchmark:spotlessJava'.
> java.lang.reflect.InvocationTargetException

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1m 29s
122 actionable tasks: 22 executed, 1 from cache, 99 up-to-date

A build scan was not published as you have not authenticated with server 'ge.opentelemetry.io'.

**Reply by trask** [reply]
try `./gradlew build` from inside of the `examples/extensions` directory

**Reply by gpalyan** [reply]
Is there anything special I need to get this extension to work. I got the example one to work and I'm trying to create a new one. I have this class. In the example you have there is an AutoService annotatation.  Is that needed with mine? Anything else needed?:

`public class CustomSdkTracerProviderConfiguration implements SdkTracerProviderConfigurer {

    @Override
    public void configure(SdkTracerProviderBuilder sdkTracerProviderBuilder) {
        sdkTracerProviderBuilder.setIdGenerator(new MyCustomIdGenerator()).build();
    }
}`

**Reply by iNikem** [reply]
Yes, SdkTracerProviderConfigurer is a [SPI](https://www.baeldung.com/java-spi), which must be "registered". You have to either use `AutoService` annotation or manually create a corresponding file in the `META-INF/services` folder
